src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleMetadata.cs

```diff
@@ -3,49 +3,215 @@
+                }
+
+                state.Current.MetadataId = reader.GetString();
+
+                // Clear the MetadataPropertyName since we are done processing Id.
+                state.Current.JsonPropertyName = default;
+
+                if (converter.ClassType == ClassType.Enumerable)
+                {
+                    // Need to Read $values property name.
+                    state.Current.ObjectState = StackFrameObjectState.MetadataValuesPropertyName;
+                }
+                else
+                {
+                    // We are done reading metadata.
+                    state.Current.ObjectState = StackFrameObjectState.MetadataPropertyValue;
+                    value = default!;
+                    return true;
+                }
             }
-            else if (metadata == MetadataPropertyName.Ref)
+
+            if (state.Current.ObjectState == StackFrameObjectState.MetadataRefPropertyEndObject)
             {
-                state.Current.ReferenceId = key;
+                if (!reader.Read())
+                {
+                    value = default!;
+                    return false;
+                }
+
+                if (reader.TokenType != JsonTokenType.EndObject)
+                {
+                    if (reader.TokenType == JsonTokenType.PropertyName)
+                    {
+                        state.Current.JsonPropertyName = GetSpan(ref reader).ToArray();
+                        ThrowHelper.ThrowJsonException_MetadataReferenceObjectCannotContainOtherProperties();
+                    }
+
+                    ThrowHelper.ThrowJsonException();
+                }
+
+                value = (T)state.Current.ReturnValue!;
+                return true;
             }
-        }
 
-        private static MetadataPropertyName GetMetadataPropertyName(ReadOnlySpan<byte> propertyName, ref ReadStack state, ref Utf8JsonReader reader)
-        {
-            Debug.Assert(state.Current.JsonClassInfo!.Options.ReferenceHandling.ShouldReadPreservedReferences());
+            if (state.Current.ObjectState == StackFrameObjectState.MetadataValuesPropertyName)
+            {
+                if (!reader.Read())
+                {
+                    value = default!;
+                    return false;
+                }
+
+                if (reader.TokenType != JsonTokenType.PropertyName)
+                {
+                    ThrowHelper.ThrowJsonException_MetadataPreservedArrayValuesNotFound(converter.TypeToConvert);
+                }
+
+                ReadOnlySpan<byte> propertyName = GetSpan(ref reader);
+
+                // Remember the property in case we get an exception.
+                state.Current.JsonPropertyName = propertyName.ToArray();
+
+                if (GetMetadataPropertyName(propertyName) != MetadataPropertyName.Values)
+                {
+                    ThrowHelper.ThrowJsonException_MetadataPreservedArrayInvalidProperty(converter.TypeToConvert, reader, ref state);
```

Comment:
This one is weird:

Before, if we had more than one `$id` property, e.g. `{"$id": "1", "$id" :"2", "$values": []}`, we would throw `The metadata property '$id' must be the first property in the JSON object.`.

Now, we are throwing `Invalid property '$id' found within a JSON object that *must only contain metadata properties* and the nested JSON array to be preserved.` which is weird since `$id` is indeed a metadata property.

Fixing this would imply to extend the logic to choose what to throw depending on `GetMetadataPropertyName`, similar to [ThrowUnexpectedMetadataException](https://github.com/dotnet/runtime/pull/2259/files#diff-761f50bce3fd5a3dc143078e59bba051R253) on line 253.

Reply:
I can't even think of how we can ensure that on a test, both messages contain the literal `'$id'` and the Json path is `$.$id` on both.

Reply:
Perhaps if we collapse the exception messages we wouldn't have to condition here?